pageserver/src/walredo.rs

```diff
@@ -587,419 +583,1147 @@ impl<C: CommandExt> CloseFileDescriptors for C {
-        wal_redo_timeout: Duration,
-    ) -> Result<Bytes, std::io::Error> {
-        // Serialize all the messages to send the WAL redo process first.
-        //
-        // This could be problematic if there are millions of records to replay,
-        // but in practice the number of records is usually so small that it doesn't
-        // matter, and it's better to keep this code simple.
+    // used to communicate hopefully utf-8 log messages
+    let stderr = child.stderr.take().expect("not taken yet");
+
+    let (result_txs, result_rx) = tokio::sync::mpsc::channel(expected_inflight);
+
+    let stdin_task = stdin_task(stdin, work_tx, work_rx, result_txs, initial, feedback);
+
+    let stdin_task = async {
+        tokio::pin!(stdin_task);
+
+        // because stdin and stdout are interconnected with a channel to faciliate
+        // pipelining, we don't need to cancel all of the tasks, just the "feeder" task.
         //
-        // Most requests start with a before-image with BLCKSZ bytes, followed by
-        // by some other WAL records. Start with a buffer that can hold that
-        // comfortably.
-        let mut writebuf: Vec<u8> = Vec::with_capacity((BLCKSZ as usize) * 3);
-        build_begin_redo_for_block_msg(tag, &mut writebuf);
-        if let Some(img) = base_img {
-            build_push_page_msg(tag, &img, &mut writebuf);
+        // using tokio::try_join! later we get the wanted "channel close or shutdown takes
+        // down all tasks after finishing the in-flight work"
+        tokio::select! {
+            ret = &mut stdin_task => ret,
+            _ = cancel.cancelled() => {
+                // instead of cancelling all of the tasks at once by:
+                //
+                // Err("tenant shutdown, cancelling walredo work")
+                //
+                // just "close" channel connecting stdin and stdout tasks so that
+                // whatever got through, will be processed.
+                //
+                // walredo process might end up seeing one more request we will not
+                // read the response to, but it's most likely killed at that point
+                // anyways.
+                Ok(())
+            }
         }
-        for (lsn, rec) in records.iter() {
-            if let NeonWalRecord::Postgres {
-                will_init: _,
-                rec: postgres_rec,
-            } = rec
-            {
-                build_apply_record_msg(*lsn, postgres_rec, &mut writebuf);
-            } else {
-                return Err(Error::new(
-                    ErrorKind::Other,
-                    "tried to pass neon wal record to postgres WAL redo",
-                ));
+    };
+
+    let stdout_task = stdout_task(stdout, result_rx);
+
+    let stderr_task = stderr_task(stderr);
+
+    async move {
+        // changed during splitting this up: run all in current task.
+        let reason = tokio::try_join!(stdin_task, stdout_task, stderr_task);
```

Comment:
Use of try_join! to take down all tasks forces a weird "Ok error" to be returned from stdout task when everything is done, will change this.

Reply:
Fixed, mothballed already into the one large commit.